### PR TITLE
Move all but resources to ir output

### PR DESCRIPTION
**Pull request by iph**
Semantics as an idea has been trasnformed to ir. All of the weird
off-by-one functionality that was there is now all in ir as a single
unit.

IR is now the "I see a parse tree and can make a lot of decisions
from it". It does the semantic-level business logic of combining
multiple areas of the parse tree to make a sensible, eventual output.

Most of parser was also carved out for formatting and output. That was
put into synthesizer instead. This feels more like a compiler now
with specific, and separate responsibilities and phases.

### CODE REVIEW

File: src/synthesizer/mod.rs [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+use crate::ir::CloudformationProgramIr;
+use crate::parser::lookup_table::MappingInnerValue;
+use crate::parser::resource::ResourceValue;
+use crate::parser::sub::{sub_parse_tree, SubValue};
+use crate::CloudformationParseTree;
+use std::collections::HashMap;
+use voca_rs::case::camel_case;
+
+// TODO - remove parse_tree
+pub fn synthesize(parse_tree: CloudformationParseTree, ir: CloudformationProgramIr) {
```
Comment (roblaks) [inline src/synthesizer/mod.rs:14]: Can we rename this file to typescript_synthesizer?

File: src/synthesizer/mod.rs [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+        println!(
+            "import * as {} from '{}';",
+            import.name,
+            import.path.join("/")
+        )
+    }
+
+    println!("export interface NoctStackProps extends cdk.StackProps {{");
+    for param in ir.constructor.inputs {
+        println!(
```
Comment (roblaks) [inline src/synthesizer/mod.rs:25]: For later, but wondering if we should have some type of indent helper that tracks how many indents to put at the start of a line.

**Review by roblaks — APPROVED** [review]

File: src/synthesizer/mod.rs (hunk 2 above)
Comment (iph) [reply src/synthesizer/mod.rs:25]: that would be nice. As of right now i just run programs like:

`cargo run example.json | tsfmt --stdin > example.ts`

Which works just as well. Will think about this in future iterations though.

File: src/synthesizer/mod.rs (hunk 1 above)
Comment (iph) [reply src/synthesizer/mod.rs:14]: Sure!

**Review by roblaks — APPROVED** [review]

**Pull request merged**